### DISCUSSION TITLE
Is it possible to import .data theme files from Turing application?

### DISCUSSION THREAD

**Discussion by YasinAl** [mathoudebine/turing-smart-screen-python#147] [Q&A] [answered]
It would be cool to be able to import the .data theme files from the Turing application. There is a certain theme that I like but I don't trust the application running on my pc 😓

As far as I can tell, the .data files are binary files. If you have any idea how to integrate this and don't have the time, maybe point me in the right direction and I can give it a shot.

**Comment by mathoudebine** [comment]
Hi! This would be a great feature indeed!
For now i have no idea how to do it because we don't know how the .data is encoded.

If you have some spare time, you can send me your .data and i can extract the background image at least.
Then you will have to create a theme.yaml from start for this project using the extracted background, and position the elements one by one manually over the background

**Reply by YasinAl** [reply]
Hi,
That is exactly what I thought too.

If you know how to extract the background image that will be great. I will do the .yaml file and put it in the themes section. Here is the link [https://drive.google.com/file/d/1kUY5u_OnDsuMvnpQXuunueL8MfFnzwIc/view?usp=sharing](url) (I can't share it directly here so I put it on my drive)

**Reply by mathoudebine** [reply]
Here you go: the background, and also a preview of what the theme looks like.
[DragonBall_theme_resources.zip](https://github.com/mathoudebine/turing-smart-screen-python/files/10684611/DragonBall_theme_resources.zip)
You will use the background to create your theme, the preview is only to remind you what the them originally looks like.

To create your theme:
- create a folder in `res/themes/` e.g. `res/themes/DragonBall`
- in this folder copy the `background.png` and a `theme.yaml` from another landscape theme, like this one: https://github.com/mathoudebine/turing-smart-screen-python/blob/main/res/themes/Landscape6Grid/theme.yaml
- follow the instructions from this page https://github.com/mathoudebine/turing-smart-screen-python/wiki/System-monitor-%3A-themes to use the Theme Editor with your theme

Have fun!

**Reply by YasinAl** [reply]
Thanks a lot. I will get it done some time this week and post it when it is ready!

**Reply by YasinAl** [reply]
Here is the first draft: https://github.com/mathoudebine/turing-smart-screen-python/discussions/156

If you have any suggestions for improvements let me know!

**Comment by mathoudebine** [comment, marked as answer]
I developed a tool called `turing-theme-extractor.py` to extract all resources (PNGs) from a theme .data file made for Turing official Windows app.
Check this wiki page to use it and create your own theme with the extracted bitmaps: [Import a theme from Turing / Xuanfang Windows app](https://github.com/mathoudebine/turing-smart-screen-python/wiki/System-monitor-%3A-themes#import-a-theme-from-turing--xuanfang-windows-app)

**Reply by YasinAl** [reply]
> I developed a tool called `turing-theme-extractor.py` to extract all resources (PNGs) from a theme .data file made for Turing official Windows app. Check this wiki page to use it and create your own theme with the extracted bitmaps: [Import a theme from Turing / Xuanfang Windows app](https://github.com/mathoudebine/turing-smart-screen-python/wiki/System-monitor-%3A-themes#import-a-theme-from-turing--xuanfang-windows-app)

The latest release breaks the custom themes that were extracted for some reason. It seems to be the assers inserted about image size. If I comment them out, they work perfectly fine!
The DragonBall theme that I shared before gives this error:
```
10:42:45 [INFO] HW revision A does not support backplate LED color setting
10:42:45 [DEBUG] Drawing Image: BACKGROUND
10:42:45 [DEBUG] Drawing Text: Network
10:42:45 [DEBUG] Drawing Text: UpArrow
10:42:45 [DEBUG] Drawing Text: DownArrow
10:42:45 [DEBUG] Drawing Text: Blank
Traceback (most recent call last):
  File "D:\turing-smart-screen-python\main.py", line 161, in <module>
    display.display_static_text()
  File "D:\turing-smart-screen-python\library\display.py", line 113, in display_static_text
    self.lcd.DisplayText(
  File "D:\turing-smart-screen-python\library\lcd\lcd_comm.py", line 223, in DisplayText
    self.DisplayPILImage(text_image, x, y)
  File "D:\turing-smart-screen-python\library\lcd\lcd_comm_rev_a.py", line 157, in DisplayPILImage
    assert image_height > 0, 'Image width must be > 0'
           ^^^^^^^^^^^^^^^^
AssertionError: Image width must be > 0
```

In Theme Editor:

```
D:\turing-smart-screen-python>py theme-editor.py DragonBall
Traceback (most recent call last):
  File "D:\turing-smart-screen-python\theme-editor.py", line 107, in <module>
    refresh_theme()
  File "D:\turing-smart-screen-python\theme-editor.py", line 79, in refresh_theme
    display.display_static_text()
  File "D:\turing-smart-screen-python\library\display.py", line 113, in display_static_text
    self.lcd.DisplayText(
  File "D:\turing-smart-screen-python\library\lcd\lcd_comm.py", line 223, in DisplayText
    self.DisplayPILImage(text_image, x, y)
  File "D:\turing-smart-screen-python\library\lcd\lcd_simulated.py", line 126, in DisplayPILImage
    assert image_height > 0, 'Image width must be > 0'
           ^^^^^^^^^^^^^^^^
AssertionError: Image width must be > 0
```

**Reply by mathoudebine** [reply]
I opened an issue for it: #170
